testdata/p4_16_samples/checksum-l4-bmv2.p4

```diff
@@ -0,0 +1,313 @@
+            0x0800: parse_ipv4;
+            default: accept;
+        }
+    }
+    state parse_ipv4 {
+        // The 4-bit IHL field of the IPv4 base header is the number
+        // of 32-bit words in the entire IPv4 header.  It is an error
+        // for it to be less than 5.  There are only IPv4 options
+        // present if the value is at least 6.  The length of the IPv4
+        // options alone, without the 20-byte base header, is thus ((4
+        // * ihl) - 20) bytes, or 8 times that many bits.
+        pkt.extract(hdr.ipv4,
+                    (bit<32>)
+                    (8 *
+                     (4 * (bit<9>) (pkt.lookahead<IPv4_up_to_ihl_only_h >().ihl)
+                      - 20)));
+        verify(hdr.ipv4.version == 4w4, error.IPv4IncorrectVersion);
+        verify(hdr.ipv4.ihl >= 4w5, error.IPv4HeaderTooShort);
+        meta.l4Len = hdr.ipv4.totalLen - (bit<16>)(hdr.ipv4.ihl)*4;
+        transition select (hdr.ipv4.protocol) {
+            6: parse_tcp;
+            17: parse_udp;
+            default: accept;
+        }
+    }
+    state parse_tcp {
+        pkt.extract(hdr.tcp);
+        transition accept;
+    }
+    state parse_udp {
+        pkt.extract(hdr.udp);
+        transition accept;
+    }
+
+}
+
+control cIngress(inout headers hdr,
+                 inout metadata meta,
+                 inout standard_metadata_t stdmeta)
+{
+    action foot() {
+        hdr.tcp.srcPort = hdr.tcp.srcPort + 1;
+        hdr.ipv4.ttl = hdr.ipv4.ttl - 1;
+        hdr.ipv4.dstAddr = hdr.ipv4.dstAddr + 4;
+    }
+    action foou() {
+        hdr.udp.srcPort = hdr.udp.srcPort + 1;
+        hdr.ipv4.ttl = hdr.ipv4.ttl - 1;
+        hdr.ipv4.dstAddr = hdr.ipv4.dstAddr + 4;
+    }
+    table guh {
+        key = {
+            hdr.tcp.dstPort : exact;
+        }
+        actions = { foot; NoAction; }
+        const default_action = NoAction;
+
+        const entries = {
+            (16w80) : foot();
+        }
+    }
+    table huh {
+        key = {
+            hdr.udp.dstPort : exact;
+        }
+        actions = { foou; NoAction; }
+        const default_action = NoAction;
+
+        const entries = {
+            (16w80) : foou();
+        }
+    }
+    apply {
+        if (hdr.tcp.isValid()) {
+            guh.apply();
+        }
+        if (hdr.udp.isValid()) {
+            huh.apply();
+        }
+    }
+}
+
+control cEgress(inout headers hdr,
+                inout metadata meta,
+                inout standard_metadata_t stdmeta)
+{
+    apply {
+    }
+}
+
+control vc(inout headers hdr,
+           inout metadata meta)
+{
+    apply {
+        // There is code similar to this in Github repo p4lang/p4c in
+        // file testdata/p4_16_samples/flowlet_switching-bmv2.p4
+        // However in that file it is only for a fixed length IPv4
+        // header with no options.  When I try to do this, it gives an
+        // error for having a varbit<> element in the tuple.
```

Comment:
Is this last sentence of the comment correct?  I do not understand it, given that hdr.ipv4.options appears in the list given as input to verify_checksum below.

Reply:
I will remove all comments in next checkin.